### PR TITLE
Support learn the best direction for missing value

### PR DESCRIPTION
**Pull request by guolinke**
Refer to #122 

will try following three treatments for missing values, and auto choose the best one.
1. min(feature_values)
2. 0
3. max(feature_values)

How to use: 
1. The zero feature values (```-1e-20f < value <= 1e-20f```) will be treated as missing values,
2. If you want to keep both zero values and missing values, please convert zero values manually ( absolute value should greater than ```1e-20f```) .

### CODE REVIEW

**Comment by guolinke** [comment]
@Laurae2 can you re-benchmark base on this ?

**Comment by Laurae2** [comment]
@guolinke I can try when I get time on my servers. I think you can expect some preliminary results during the weekend.

btw, xgboost taking half the time now for my benchmarks (not ended yet but most are very fast). I have to check later if LightGBM is slower now. I think only LightGBM v1 is potentially faster. v2 is slower.

**Comment by rgranvil** [comment]
@guolinke I was testing the new handling of missing values.  I could create test cases where missing was treated as 0 and as max(feature value), but not min(feature value).  

In this example I have a single predictor x and a label y.  I fit a single tree LightGBM model and append the model predictions.  Based on the description, it seems like NA should be treated as min(feature) but it appears to be treated as 0.  

Data:

x | y | prediction
--|--|-------------
**NA** | **1** | **0.1192029**
-9 | 1 | 0.7913915
-8 | 1 | 0.7913915
-7 | 1 | 0.7913915
-6 | 1 | 0.7913915
-5 | 0 | 0.7913915
-4 | 0 | 0.1192029
-3 | 0 | 0.1192029
-2 | 0 | 0.1192029
-1 | 0 | 0.1192029


**lgb.dump:**
"name":"tree",
"num_class":1,
"num_tree_per_iteration":1,
"label_index":0,
"max_feature_idx":0,
"feature_names":["x"],
"tree_info":[{"tree_index":0,"num_leaves":2,
"shrinkage":1,
"has_categorical":0,
"tree_structure":{
"split_index":0,
"split_feature":0,
"split_gain":6.6666666666666607,
"threshold":-4.5,
"decision_type":"no_greater",
**"default_value":2.2250738585072014e-308,**
"internal_value":0,
"internal_count":10,
"left_child":{
"leaf_index":0,
"leaf_parent":0,
"leaf_value":1.3333333333333324,
"leaf_count":6

**Code:**
```r
library(lightgbm)

df <- data.frame(x = as.numeric(c(NA,-9:-1)), y = c(rep(1, 5), rep(0, 5)))
lgb.data <- lgb.Dataset(
  data = as.matrix(df["x"])
  ,label = df$y
  ,max_bin = 255
)
lgb.model <- lgb.train(
  data = lgb.data
  ,nround = 1
  ,num_leaves = 2
  ,min_data = 1
  ,learning_rate = 1
  ,obj = "binary"
)
writeLines(lgb.dump(lgb.model))
df$prediction <- predict(lgb.model, as.matrix(df))
df

```

**Comment by guolinke** [comment]
@rgranvil in your case, treating as 0 is the same as treating as min, since they have no difference when comparing with ```threshold```. (only have positive feature values.)

The only case that NA will be really treated as ```0``` is, there are negative feature values. In that case, treating as ```0``` is different with treating as min.

**Comment by guolinke** [comment]
@rgranvil 
From your model, the NA will be treated as ```2.2250738585072014e-308```, this actually is zero. And the prediction of NA is correct as well.

**Comment by rgranvil** [comment]
@guolinke I agree, but why would the missing value logic not treat NA as min(feature) in this case? The NA observation has a y value of 1, so I would think the "best" choice would be to assign NA to min(feature) with a prediction ~0.79 instead of zero with a prediction ~0.12.

**Comment by guolinke** [comment]
@rgranvil Fixed in the latest commit. It has a bug for the -inf conversion.

**Comment by rgranvil** [comment]
Thanks @guolinke.  I reran my tests and it appears to be working correctly now.

**Comment by Laurae2** [comment]
Putting some results here coming from https://github.com/Laurae2/gbt_benchmarks/issues/1

| Algorithm | Time (s) | Perf (AUC) | Best Iteration |
| --- | ---: | ---: | ---: |
| xgboost depthwise | 694.043 | 0.7194797 | 603 |
| xgboost lossguide | 696.006 | 0.7194797 | 603 |
| LightGBM master | 655.582 | 0.7180479 | 796 |
| LightGBM v2 | 675.591 | 0.7146346 | 636 |
| LightGBM v1 | 670.116 | 0.7167827 | 716 |

Speed with MinGW, not VS. There are issues with MinGW threading.

**Comment by qashqay654** [comment]
Hello. After this commit it's impossible to load old models and my trees have very strange behaviour. Is it possible to turn of default values or turn them all to zeros?

**Comment by guolinke** [comment]
@qashqay654 
you can disable it by set ```use_missing=false```. However, The old model file cannot be loaded any more.

**Comment by qashqay654** [comment]
Is it possible to disable it in python?

**Comment by guolinke** [comment]
@qashqay654 
The same. You can set it to the parameters when using lgb.train or sklearn interface.
